pkg/asset/manifests/ingress.go

```diff
@@ -37,38 +41,68 @@ func (*Ingress) Dependencies() []asset.Asset {
 	}
 }
 
-// Generate generates the ingress config and its CRD.
+// Generate generates the ingress cluster config and default ingresscontroller.
 func (ing *Ingress) Generate(dependencies asset.Parents) error {
 	installConfig := &installconfig.InstallConfig{}
 	dependencies.Get(installConfig)
 
-	config := &configv1.Ingress{
+	clusterConfig, err := ing.generateClusterConfig(installConfig.Config)
+	if err != nil {
+		return errors.Wrap(err, "failed to create cluster config")
+	}
+
+	defaultIngressController, err := ing.generateDefaultIngressController()
+	if err != nil {
+		return errors.Wrap(err, "failed to create default ingresscontroller")
+	}
+
+	ing.FileList = []*asset.File{
+		{
+			Filename: clusterIngressConfigFile,
+			Data:     clusterConfig,
+		},
+		{
+			Filename: defaultIngressControllerFile,
+			Data:     defaultIngressController,
+		},
+	}
+
+	return nil
+}
+
+func (ing *Ingress) generateClusterConfig(installConfig *types.InstallConfig) ([]byte, error) {
+	obj := &configv1.Ingress{
 		TypeMeta: metav1.TypeMeta{
-			APIVersion: configv1.SchemeGroupVersion.String(),
+			APIVersion: configv1.GroupVersion.String(),
 			Kind:       "Ingress",
 		},
 		ObjectMeta: metav1.ObjectMeta{
 			Name: "cluster",
 			// not namespaced
 		},
 		Spec: configv1.IngressSpec{
-			Domain: fmt.Sprintf("apps.%s", installConfig.Config.ClusterDomain()),
+			Domain: fmt.Sprintf("apps.%s", installConfig.ClusterDomain()),
 		},
 	}
+	return yaml.Marshal(obj)
+}
 
-	configData, err := yaml.Marshal(config)
-	if err != nil {
-		return errors.Wrapf(err, "failed to create %s manifests from InstallConfig", ing.Name())
-	}
-
-	ing.FileList = []*asset.File{
-		{
-			Filename: ingCfgFilename,
-			Data:     configData,
+func (ing *Ingress) generateDefaultIngressController() ([]byte, error) {
+	var replicas int32 = 2
+	obj := &operatorv1.IngressController{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: operatorv1.GroupVersion.String(),
+			Kind:       "IngressController",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "openshift-ingress-operator",
+			Name: "default",
+		},
+		Spec: operatorv1.IngressControllerSpec{
```

Comment:
Are you suggesting the `render` command loads the YAML produced from here, mutates it, and then re-writes it?

Reply:
I would like if we don't make any defaulting decisions in installer, and let the operator control those.

that would be my preference.

Reply:
Are you suggesting the following?

1. The ingress-operator knows how to create a default ingresscontroller if one doesn't exist.  (This is today's behavior.)
2. If the user runs `openshift-install create manifests`, the installer _will not render_ `manifests/default-ingresscontroller.yaml`, but if the user creates `manifests/default-ingresscontroller.yaml`, the installer knows how to consume and apply the resource.

cc @deads2k

Reply:
> If the user runs openshift-install create manifests, the installer will not render manifests/default-ingresscontroller.yaml, 

Ahh sorry for creating confusion, it's fine to add the default to the installer. What I meant to say was that,
- the installer has empty spec default
  decisions like hostnetwork vs LB vs service-network-only etc
  decisions like number of replicas
  are not injected into the installer go-code, rather handled if needed by the bootstrap-host-render or if possible are already defaulted by API spec.
-  the installer if needed will set `internal` to the default in go-code
   but the other defaults like ones above are still controlled by the bootstrap-host-render or if possible are already defaulted by API spec.

it that more clearer?

Reply:
You can provide an empty spec and the operator will use good defaults. That's what I'm basically trying to do here.

The only reason `.spec.replicas` is added in the installer code I'm introducing (duplicating the ingress-controller's defaulting) is a UX issue. A nil replica field breaks scale sub resource in clients, requiring the use of a complex patch call where a scale call is preferred. So encoding just the replica count here seems like a useful exception to make, although it could very well come with drawbacks I'm not foreseeing. If there are indeed some drawbacks, we would want to find another way to fill in that default pre-persistence (which sounds somewhat complicated if we're talking about post-processing the user's resource during the `render` command or something).

Reply:
> The only reason .spec.replicas is added in the installer code I'm introducing (duplicating the ingress-controller's defaulting) is a UX issue. A nil replica field breaks scale sub resource in clients, requiring the use of a complex patch call where a scale call is preferred. So encoding just the replica count here seems like a useful exception to make,

+1

Reply:
@abhinavdahiya if the installconfig is such that no ingress customizations are even provided (as in the default public case), should the installer even generate the ingresscontroller YAML at all?

Reply:
What if I just remove rendering of the default entirely for now? Users can still add their own to the manifests directory.

And then when https://github.com/openshift/installer/pull/2521 lands I can write a default only if the installconfig says the cluster is internal?